aten/src/ATen/native/RNN.cpp

```diff
@@ -365,23 +371,39 @@ struct FullBidirectionalLayer : Layer<Tensor, pair_of<dir_hidden_type>, pair_of<
 
   output_type operator()(const Tensor& input, const hidden_type& input_hidden, const param_type& params) const override {
     auto step_inputs = input.unbind(0);
-    auto fw_result = layer_(step_inputs, input_hidden.first, params.first);
-    auto fw_output = at::stack(fw_result.outputs, 0);
 
-    auto rev_step_inputs = reverse(std::move(step_inputs));
-    auto rev_result = layer_(rev_step_inputs, input_hidden.second, params.second);
-    std::reverse(rev_result.outputs.begin(), rev_result.outputs.end());
+    std::atomic<bool> task_finished {false};
+    std::mutex task_mutex;
+    std::condition_variable task_cv;
+
+    using unstacked_output_type = typename FullLayer<dir_hidden_type, cell_params>::unstacked_output_type;
+    unstacked_output_type fw_result;
+    at::Tensor fw_output;
+    at::intraop_launch([&]() {
```

Comment:
can it (optionally) return some Future-like abstraction? it's annoying to reimplement the same pattern in many places. How about ivalue::Future?

though in this case it can be just regular std::promise<void>, right?

Reply:
There's an annoying issue in ivalue::Future that we need to fix first, it forces us to use std::shared_ptr to pass a future inside lambda and I don't want to add this overhead atm

the issue happens when the future is defined locally somewhere before lambda, then passed into lambda, then we schedule lambda, then we wait on future after scheduling lambda.

At some moment 'markCompleted' is going to be called from inside the lambda. Our current 'wait' is implemented as a callback.
While future still executes markCompleted it will unblock waiter that will finish the work and return and destroy the future.

My plan was to address this separately and provide a more convenient API, but that's not blocking atm, we can simply use condition_variable here

Reply:
actually, let me check std::future/std::promise - maybe it will be possible to use it from intraop_launch

Reply:
I think we could also use std future in ParallelNative.h too

Reply:
@dzhulgakov sent https://github.com/pytorch/pytorch/pull/21764